R/conditional-transform.R

```diff
@@ -0,0 +1,150 @@
+        cases[[i]] <- evalLHS(formula, data)
+        if (!inherits(cases[[i]], "CrunchLogicalExpr")) {
+            halt("The LHS provided is not a CrunchLogicalExpr: ", formulaLHS(formula))
+        }
+        values[[i]] <- evalRHS(formula, data)
+    }
+    case_indices <- lapply(cases, which)
+    
+    # dedpulicate indices, favoring the first true condition
+    case_indices <- lapply(seq_along(case_indices), function(i) {
+        setdiff(case_indices[[i]], unlist(case_indices[seq_len(i-1)]))
+    })
+    
+    # grab the values needed from source variables
+    values_to_fill <- Map(function(ind, var) {
+        if (inherits(var, c("CrunchVariable", "CrunchExpr"))) {
+            # grab the variable contents at inds
+            return(as.vector(var[ind]))
+        } else {
+            # if var isn't a crunch variable or expression, just return var
+            return(var)
+        }
+    }, ind = case_indices, var = values)
+    
+    # setup NAs for as default
+    n_rows <- nrow(CrunchDataset(crGET(datasetReference(cases[[1]]))))
+    result <- rep(else_condition, n_rows)
+    
+    # fill values
+    for (i in seq_along(case_indices)) {
+        if (type == "numeric") {
+            trans <- as.numeric
+        } else {
+            trans <- as.character
+        }
+        vals <- trans(values_to_fill[[i]])
+        result[case_indices[[i]]] <- vals
+    }
+ 
+    if (type == "categorical") {
+        # if categories are supplied and there are any
+        if (missing(categories)) {
+            result <- factor(result)
+            # if categories aren't a Categories object,
+            # make categories from names
+            categories <- Categories(data = categoriesFromLevels(levels(result)))
+            categories <- c(categories, Category(data=.no.data))
+        } else {
+            if (!is.categories(categories)) {
+                # if categories aren't a Categories object,
+                # make categories from names
+                categories <- Categories(data = categoriesFromLevels(categories))
+                categories <- c(categories, Category(data=.no.data))
+            } else if (is.null(categories[["No data"]])) {
+                categories <- c(categories, Category(data=.no.data))
+            }
+
+            uni_results <- unique(result[!is.na(result)])
+            results_not_categories <- !uni_results %in% names(categories)
+            if (any(results_not_categories)) {
+                halt("there were categories in the results (",
+                     serialPaste(uni_results[results_not_categories]),
+                     ") that were not specified in categories")
+            }
+            result <- factor(result, levels = names(categories))
+        }
+    }
+
+    var_def$type <- type
+    if (type == "categorical") {
+        # make a category list to send with VariableDefinition and then store that and convert values to ids values
+        category_list <- listifyCategories(categories)
```

Comment:
I needed someway to go from a `Categories` object to a plain list that would then get parsed into json for the variable definition. I tried a bunch of different things, but I didn't find anything already in the package that would do this. It might be better to have this as the `as.list` method (I didn't because `Categories` objects already inherit from lists through s4 inheritance magic and that seems a bit strange to do).

Reply:
`Categories` already have a JSON serializing method (https://github.com/Crunch-io/rcrunch/blob/master/R/json.R), so you don't need to do anything special to make them turn into JSON correctly. (Note that the `categories<-` method just PATCHes with the `Categories` object.)

Reply:
It turns out that calling `toJSON()` on a `VariableDefinition` doesn't trigger the list-based `lapplying` of `jsonprep()`, which was leading to errors. I've added that, and now can fully do away with this listifying.